source/slang/ir-sccp.cpp

```diff
@@ -0,0 +1,933 @@
+        // `Constant`s.
+        SLANG_ASSERT(left.flavor == LatticeVal::Flavor::Constant);
+        //
+        SLANG_ASSERT(right.flavor == LatticeVal::Flavor::Constant);
+
+        // If the two lattice values represent the *same* constant value
+        // (they are the same singleton set) then the union is that
+        // singleton set as well.
+        //
+        // TODO: This comparison assumes that constants with
+        // the same value with be represented with the
+        // same instruction, which is not *always*
+        // guaranteed in the IR today.
+        //
+        if(left.value == right.value)
+            return left;
+
+        // Otherwise, we have two distinct singleton sets, and their
+        // union should be a set with two elements. We can't represent
+        // that on the lattice for SCCP, so the proper lower bound
+        // is the universal set (`Any`)
+        //
+        return LatticeVal::getAny();
+    }
+
+    // During the execution of the SCCP algorithm, we will track our best
+    // "estimate" so far of the set of values each instruction could take
+    // on. This amounts to a mapping from IR instructions to lattice values,
+    // where any instruction not present in the map is assumed to default
+    // to the `None` case (the empty set)
+    //
+    Dictionary<IRInst*, LatticeVal> mapInstToLatticeVal;
+
+    // Updating the lattice value for an instruction is easy, but we'll
+    // use a simple function to make our intention clear.
+    //
+    void setLatticeVal(IRInst* inst, LatticeVal const& val)
+    {
+        mapInstToLatticeVal[inst] = val;
+    }
+
+    // Querying the lattice value for an instruction isn't *just* a matter
+    // of looking it up in the dictionary, because we need to account for
+    // cases of lattice values that might come from outside the current
+    // function.
+    //
+    LatticeVal getLatticeVal(IRInst* inst)
+    {
+        // Instructions that represent constant values should always
+        // have a lattice value that reflects this.
+        //
+        switch( inst->op )
+        {
+        case kIROp_IntLit:
+        case kIROp_FloatLit:
+        case kIROp_StringLit:
+        case kIROp_BoolLit:
+            return LatticeVal::getConstant(inst);
+            break;
+
+        // TODO: We might want to start having support for constant
+        // values of aggregate types (e.g., a `makeArray` or `makeStruct`
+        // where all the operands are constant is itself a constant).
+
+        default:
+            break;
+        }
+
+        // We might be asked for the lattice value of an instruction
+        // not contained in the current function. When that happens,
+        // we will treat it as having potentially any value, rather
+        // than the default of none.
+        //
+        auto parentBlock = as<IRBlock>(inst->getParent());
+        if(!parentBlock || parentBlock->getParent() != code) return LatticeVal::getAny();
```

Comment:
Pedantic point, but don't you want the return tabbed on the next line - as seen below.